google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/stream.rb

```diff
@@ -0,0 +1,348 @@
+                    @inventory.add rec_msg.ack_id
+
+                    perform_callback_async rec_msg
+                  end
+                end
+                synchronize { pause_streaming! }
+              rescue StopIteration
+                break
+              end
+            end
+          rescue GRPC::DeadlineExceeded
+            # The GRPC client will raise when stream is opened longer than the
+            # timeout value it is configured for. When this happends, restart the
+            # stream stealthly.
+            synchronize { start_streaming! }
+          rescue => e
+            synchronize do
+              @request_queue.unshift Google::Cloud::Error.from_error(e)
+            end
+          end
+
+          # rubocop:enable all
+
+          def perform_callback_async rec_msg
+            Concurrent::Future.new(executor: subscriber.thread_pool) do
+              subscriber.callback.call rec_msg
+            end.execute
+          end
+
+          def start_streaming!
+            # signal to the previous queue to shut down
+            @background_thread.kill if @background_thread
+            @request_queue.push self if @request_queue
+
+            @request_queue = EnumeratorQueue.new self
+            @request_queue.push initial_input_request
+            @output_enum = subscriber.service.streaming_pull \
+              @request_queue.each_item
+
+            @background_thread = Thread.new { background_run }
+          end
+
+          def pause_streaming!
+            return unless pause_streaming?
+
+            @paused = true
+          end
+
+          def pause_streaming?
+            return if @paused
+
+            @inventory.full?
+          end
+
+          def unpause_streaming!
+            return unless unpause_streaming?
+
+            @paused = nil
+            # Need to recreate the enums otherwise we get the error:
+            # "fiber called across threads"
+            start_streaming!
+          end
+
+          def unpause_streaming?
+            return if @paused.nil?
+
+            @inventory.count < @inventory.limit*0.8
+          end
+
+          def initial_input_request
+            Google::Pubsub::V1::StreamingPullRequest.new.tap do |req|
+              req.subscription = subscriber.subscription_name
+              req.stream_ack_deadline_seconds = subscriber.deadline
+              req.modify_deadline_ack_ids += @inventory.ack_ids
+              req.modify_deadline_seconds += \
+                @inventory.ack_ids.map { subscriber.deadline }
+            end
+          end
+
+          ##
+          # Makes sure the values are the `ack_id`. If given several
+          # {ReceivedMessage} objects extract the `ack_id` values.
+          def coerce_ack_ids messages
+            Array(messages).flatten.map do |msg|
+              msg.respond_to?(:ack_id) ? msg.ack_id : msg.to_s
+            end
+          end
+
+          # @private
+          class Inventory
```

Comment:
I understand that, but I think it's more confusing for users, that don't want to think about streams, just their total memory limits, and it's different than we've done this in Java/C# where the number of streams is invisible to the user. I'm happy for the number to be user configurable, but I think that flow control limits should be per subscriber.